staging/src/k8s.io/cloud-provider/node/helpers/status.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package helpers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// PatchNodeStatus patches node status.
+func PatchNodeStatus(c v1core.CoreV1Interface, nodeName types.NodeName, oldNode *v1.Node, newNode *v1.Node) (*v1.Node, []byte, error) {
```

Comment:
I'm usually for duplicating small bits of code to remove dependencies but I'm not sure about this one. The patch logic for node addresses has been problematic in the past (see https://github.com/kubernetes/kubernetes/pull/79391) and I think we should try to make the logic consistent across kubelet and CCM. Wondering if there's a way to remove that internal dep without duplicating the implementation. Can kubelet import this implementation or can we find a different home for this?

Reply:
I should’ve mentioned this is just a WIP to kick-off discussions around this one.
That said,I agree se should fine a new home for this. For now `k8s.io/kubelet` is mostly for component config types, `k8s.io/cloud-provider` doesn’t seem like the right place. The only one that seems fine is `k8s.io/component-base`, thoughts @liggitt

Reply:
> The only one that seems fine IMO is `k8s.io/component-base`

I would not expect component-base to hold helpers for specific components.

Reply:
@liggitt - the only other solution is to move it to `k8s.io/cloud-provider` and import `k8s.io/cloud-provider/node/helpers` in:

- the node_controller to update node addresses
- the kubelet
- the attach_detach_controller
- csi

Reply:
@liggitt - the other solution is to move it to `k8s.io/cloud-provider` and import `k8s.io/cloud-provider/node/helpers` in:

- the node_controller to update node addresses
- the kubelet
- the attach_detach_controller
- csi

Reply:
ping @liggitt @andrewsykim